gossip/src/push_active_set.rs

```diff
@@ -1,21 +1,116 @@
 use {
-    crate::weighted_shuffle::WeightedShuffle,
+    crate::{
+        cluster_info::REFRESH_PUSH_ACTIVE_SET_INTERVAL_MS,
+        low_pass_filter as lpf,
+        stake_weighting_config::{get_gossip_config_from_account, WeightingConfig},
+        weighted_shuffle::WeightedShuffle,
+    },
     indexmap::IndexMap,
+    log::error,
     rand::Rng,
     solana_bloom::bloom::{Bloom, ConcurrentBloom},
     solana_native_token::LAMPORTS_PER_SOL,
     solana_pubkey::Pubkey,
+    solana_runtime::bank::Bank,
     std::collections::HashMap,
 };
 
 const NUM_PUSH_ACTIVE_SET_ENTRIES: usize = 25;
+const ALPHA_MIN: u64 = lpf::SCALE.get();
+const ALPHA_MAX: u64 = 2 * lpf::SCALE.get();
+const DEFAULT_ALPHA: u64 = ALPHA_MAX;
+// Low pass filter convergence time (ms)
+const DEFAULT_TC_MS: u64 = 30_000;
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum WeightingMode {
+    // alpha = 2.0 -> Quadratic
+    Static,
+    // alpha in [1.0, 2.0], smoothed over time, scaled up by 1000 to avoid floating-point math
+    Dynamic,
+}
+
+#[inline]
+fn get_weight(bucket: u64, alpha: u64) -> u64 {
+    debug_assert!((ALPHA_MIN..=ALPHA_MIN + lpf::SCALE.get()).contains(&alpha));
+    let b = bucket + 1;
+    lpf::interpolate(b, alpha.saturating_sub(ALPHA_MIN))
+}
 
 // Each entry corresponds to a stake bucket for
 //     min stake of { this node, crds value owner }
 // The entry represents set of gossip nodes to actively
 // push to for crds values belonging to the bucket.
-#[derive(Default)]
-pub(crate) struct PushActiveSet([PushActiveSetEntry; NUM_PUSH_ACTIVE_SET_ENTRIES]);
+pub(crate) struct PushActiveSet {
+    entries: [PushActiveSetEntry; NUM_PUSH_ACTIVE_SET_ENTRIES],
+    alpha: u64, // current alpha (fixed-point, 1000–2000)
+    mode: WeightingMode,
+    filter_k: u64,
+    tc_ms: u64,
+}
+
+impl Default for PushActiveSet {
```

Comment:
`Default` is usually bug. prefer literal instantiation. add a `fn default_for_tests()` under dcou if that's where the "convenience" of `Default` is desired